src/zino/statemodels.py

```diff
@@ -363,3 +368,156 @@ class AlarmEvent(Event):
     @property
     def subindex(self) -> SubIndex:
         return self.alarm_type
+
+
+class MatchType(StrEnum):
+    """The set of allowable BGP oper states"""
+
+    REGEXP = "regexp"
+    STR = "str"
+    EXACT = "exact"
+    INTF_REGEXP = "intf-regexp"
+
+
+class PmType(StrEnum):
+    PORTSTATE = "portstate"
+    DEVICE = "device"
+
+
+class PlannedMaintenance(BaseModel):
+    id: Optional[int] = None
+    start_time: datetime.datetime
+    end_time: datetime.datetime
+    type: PmType
+    match_type: MatchType
+    match_device: Optional[str]
+    match_expression: str
+    log: List[LogEntry] = []
+    event_ids: List[int] = []
+
+    def start(self, state: "ZinoState"):
+        events = self._get_or_create_events(state)
+        for event in events:
+            # get special handling of embryonic -> open transition first
+            state.events.commit(event)
+            event.state = EventState.IGNORED
+            state.events.commit(event)
+            self.event_ids.append(event.id)
+
+    def end(self, state: "ZinoState"):
+        for event_id in self.event_ids:
+            event = state.events.checkout(event_id)
+            event.state = EventState.OPEN
+            state.events.commit(event)
+
+    def add_log(self, message: str) -> LogEntry:
+        entry = LogEntry(message=message)
+        self.log.append(entry)
+        return entry
+
+    def matches_event(self, event: Event, state: "ZinoState") -> bool:
+        """Returns true if `event` will be affected by this planned maintenance"""
+        raise NotImplementedError
+
+    def _get_or_create_events(self, state: "ZinoState") -> list[Event]:
+        """Creates/gets events that are affected by the given starting planned
+        maintenance
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _string_match(pattern: str, string: str) -> bool:
+        """This should behave like tcl string match https://wiki.tcl-lang.org/page/string+match
+        Returns true if `string` matches `pattern`.
+        """
+        return fnmatch.fnmatch(string, pattern)
+
+    @staticmethod
+    def _regex_match(pattern: str, string: str) -> bool:
+        """Matches `string` against regex expression `pattern`.
+        Returns true if there is a match.
+        """
+        compiled_pattern = re.compile(pattern)
+        return bool(compiled_pattern.match(string))
```

Comment:
Compiling the pattern before using it here makes no sense.  The compiling step makes sense if we want to optimize performance by reusing the compiled pattern over and over, but here it's essentially throwing away the compiled pattern after each use.

Either use `re.match()` directly, or ensure that the PM classes pre-compile any regex-patterns so they can be re-used.

Reply:
fixed